sale_order_lot_selection/models/sale.py

```diff
@@ -52,32 +52,28 @@ def get_move_from_line(self, line):
                         raise UserError(_('Can\'t retrieve lot on stock'))
         return move
 
-    @api.model
-    def _check_move_state(self, line):
-        if line.lot_id:
-            move = self.get_move_from_line(line)
-            if move.state == 'confirmed':
-                move._action_assign()
-                move.refresh()
-            if move.state != 'assigned':
-                raise UserError(_('Can\'t reserve products for lot %s') %
-                                line.lot_id.name)
+    def forbid_unavailable_lot_reservation(self):
+        """ You may want use unavailable lot in your stock move:
+            i.e. you sell a product manufactured by a supplier
+            but with your requirements including traceability on received lot.
+
+            In this case override this method to return True in your use case
+        """
         return True
 
-    @api.multi
     def action_confirm(self):
-        res = super(SaleOrder, self.with_context(sol_lot_id=True))\
-            .action_confirm()
+        res = super().action_confirm()
         for line in self.order_line:
             if line.lot_id:
-                unreserved_moves = line.move_ids.filtered(
-                    lambda move: move.product_uom_qty !=
-                    move.reserved_availability
-                )
-                if unreserved_moves:
+                move = self.get_move_from_line(line)
+                if move.state == 'confirmed':
+                    move._action_assign()
+                    move.refresh()
+                if (
+                    move.state != "assigned"
+                    and self.forbid_unavailable_lot_reservation()
```

Comment:
would be better to pass the line and/or the move to this forbid_unavailable_lot_reservation hook (so we can have the lot and the product), so you can do the necessary checking that you are indeed in your special case, no?